composer/core/state.py

```diff
@@ -504,18 +546,38 @@ def fsdp_enabled(self):
                 return True
         return False
 
+    def dataset_state_dict(self) -> Dict[str, Any]:
+        obj = {
+            'train': None,
+            'eval': {},
+        }
+
+        dataset = self._dataset_of(self.train_dataloader)
+        if hasattr(dataset, 'state_dict'):
+            obj['train'] = dataset.state_dict()
+
+        for evaluator in self.evaluators:
+            dataset = self._dataset_of(evaluator.dataloader)
+            if hasattr(dataset, 'state_dict'):
+                st = dataset.state_dict()
+                st['sessions'] = []  # Don't save eval split's sessions, because we do not checkpoint during eval.
+                obj['eval'][evaluator.label] = st
```

Comment:
```suggestion
                obj['eval'][evaluator.label] = dataset.state_dict()
```
should be fine. When you call `dataset.state_dict()`, it'll never be in the middle of epoch

Reply:
Something about switching to pre-incrementing epoch made this ugly